pkg/k8s/watchers/cilium_endpoint_slice_subscriber.go

```diff
@@ -123,51 +133,112 @@ func (cs *cesSubscriber) OnUpdate(oldCES, newCES *cilium_v2a1.CiliumEndpointSlic
 // and calls endpointDeleted only for remoteNode CEPs.
 func (cs *cesSubscriber) OnDelete(ces *cilium_v2a1.CiliumEndpointSlice) {
 	for i, ep := range ces.Endpoints {
+		CEPName := ces.Namespace + "/" + ep.Name
 		log.WithFields(logrus.Fields{
 			"CESName": ces.GetName(),
-			"CEPName": ep.Name,
+			"CEPName": CEPName,
 		}).Debug("CES deleted, calling endpointDeleted")
 		c := k8s.ConvertCoreCiliumEndpointToTypesCiliumEndpoint(&ces.Endpoints[i], ces.Namespace)
 		// LocalNode already deleted the CEP.
 		// Hence, skip processing endpointDeleted for localNode CEPs.
-		if p := cs.kWatcher.endpointManager.LookupPodName(k8sUtils.GetObjNamespaceName(c)); p != nil {
+		if p := cs.epCache.LookupPodName(k8sUtils.GetObjNamespaceName(c)); p != nil {
 			continue
 		}
 		// Delete CEP if and only if that CEP is owned by a CES, that was used during CES updated.
 		// Delete CEP only if there is match in CEPToCES map and also delete CEPName in CEPToCES map.
-		if cesName := cepMap.getCESName(ces.Namespace + "/" + ep.Name); cesName == ces.GetName() {
-			cs.kWatcher.endpointDeleted(c)
-			cepMap.deleteCEP(ep.Name)
+		cs.deleteCEPfromCES(CEPName, ces.GetName(), c)
+	}
+}
+
+// deleteCEP deletes the CEP and CES from the map.
+// If this was last CES for the CEP it triggers endpointDeleted.
+// If this was used CES for the CEP it picks other CES and triggers endpointUpdated.
+func (cs *cesSubscriber) deleteCEPfromCES(CEPName, CESName string, c *types.CiliumEndpoint) {
+	needUpdate := cs.cepMap.currentCES[CEPName] == CESName
+	cs.cepMap.deleteCEP(CEPName, CESName)
+	if needUpdate {
+		cep, exists := cs.cepMap.getCEP(CEPName)
+		if !exists {
+			log.WithFields(logrus.Fields{
+				"CESName": CESName,
+				"CEPName": CEPName,
+			}).Info("CEP deleted, calling endpointDeleted")
+			cs.epWatcher.endpointDeleted(c)
+		} else {
+			log.WithFields(logrus.Fields{
+				"CESName": CESName,
+				"CEPName": CEPName,
+			}).Info("CEP deleted, other CEP exists, calling endpointUpdated")
+			cs.epWatcher.endpointUpdated(c, cep)
 		}
 	}
 }
 
-// cepToCESmap is used to map CiliumEndpoint name to CiliumEndpointBatch name.
+// addCEPwithCES insert CEP with CES ot the map and triggers endpointUpdated.
+func (cs *cesSubscriber) addCEPwithCES(CEPName, CESName string, newC *types.CiliumEndpoint) {
+	oldC, _ := cs.cepMap.getCEP(CEPName)
+	cs.cepMap.insertCEP(CEPName, CESName, newC)
+	cs.epWatcher.endpointUpdated(oldC, newC)
+}
+
+// / cepToCESmap is used to map CiliumEndpoint name to CiliumEndpointSlice names.
+// In steady state, there should be exactly one CiliumEndpointSlice associated
+// with a CiliumEndpoint. But when a CEP is being transferred between two CESes,
+// there will be a brief period of time in which the CEP exists in both the CESes.
 type cepToCESmap struct {
 	cesMutex lock.RWMutex
```

Comment:
When is `cesMutex` used?
Can you add a comment here about it?

Reply:
Done, PTAL

Reply:
I don't see the changes. Perhaps you didn't push them?

Reply:
pushed now

Reply:
lgtm

Reply:
Also, this could just be a regular lock.

Reply:
Changed to Mutex